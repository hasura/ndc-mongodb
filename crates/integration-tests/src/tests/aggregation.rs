use insta::assert_yaml_snapshot;
use serde_json::json;

use crate::graphql_query;

#[tokio::test]
async fn runs_aggregation_over_top_level_fields() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query($albumId: Int!) {
                  track(order_by: { id: Asc }, where: { albumId: { _eq: $albumId } }) {
                    milliseconds
                    unitPrice
                  }
                  trackAggregate(
                    filter_input: { order_by: { id: Asc }, where: { albumId: { _eq: $albumId } } }
                  ) {
                    _count
                    milliseconds {
                      _avg
                      _max
                      _min
                      _sum
                    }
                    unitPrice {
                      _count
                      _count_distinct
                    }
                  }
                }
            "#
        )
        .variables(json!({ "albumId": 9 }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_extended_json_representing_mixture_of_numeric_types() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This doesn't affect native queries that don't use the $documents stage.
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      _avg
                      _count
                      _max
                      _min
                      _sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "decimal|double|int|long" }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_mixture_of_numeric_and_null_values() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This doesn't affect native queries that don't use the $documents stage.
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      _avg
                      _count
                      _max
                      _min
                      _sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "double|null" }))
        .run()
        .await?
    );
    Ok(())
}
