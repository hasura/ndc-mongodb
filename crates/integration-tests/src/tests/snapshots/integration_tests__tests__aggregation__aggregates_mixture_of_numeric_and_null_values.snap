---
source: crates/integration-tests/src/tests/aggregation.rs
expression: "graphql_query(r#\"\n                query ($types: String!) {\n                  extendedJsonTestDataAggregate(\n                    filter_input: { where: { type: { _regex: $types } } }\n                  ) {\n                    value {\n                      _avg\n                      _count\n                      _max\n                      _min\n                      _sum\n                      _count_distinct\n                    }\n                  }\n                  extendedJsonTestData(where: { type: { _regex: $types } }) {\n                    type\n                    value\n                  }\n                }\n            \"#).variables(json!({\n                        \"types\": \"double|null\"\n                    })).run().await?"
---
data:
  extendedJsonTestDataAggregate:
    value:
      _avg:
        $numberDouble: "3.5"
      _count: 2
      _max:
        $numberDouble: "4.0"
      _min:
        $numberDouble: "3.0"
      _sum:
        $numberDouble: "7.0"
      _count_distinct: 2
  extendedJsonTestData:
    - type: double
      value:
        $numberDouble: "3.0"
    - type: double
      value:
        $numberDouble: "4.0"
    - type: "null"
      value: ~
errors: ~
